### DISCUSSION TITLE
Help with "[webpack.cache.PackFileCacheStrategy] Serializing big strings (834kiB) impacts deserialization performance (consider using Buffer instead and decode when needed)"

### DISCUSSION THREAD

**Discussion by lancejpollard** [vercel/next.js#57843] [Help]
### Summary

I keep getting this in my Next.js web app:

> info Creating an optimized production build .<w> [webpack.cache.PackFileCacheStrategy] Serializing big strings (834kiB) impacts deserialization performance (consider using Buffer instead and decode when needed)

I get this when deploying to Vercel / production, and when running `next build` locally. How do I debug this to figure out what is causing this statement? It is causing my builds to take over 5 minutes each time, which is annoying to work with. Is there anything straightforward to either:

1. Get rid of this problem entirely?
2. Log some stuff so I can see what is even _causing_ this problem, so I can decide if I should change something?

Thank you for your help in advance!

### Additional information

Here is my basic info:

```
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 22.5.0
Binaries:
  Node: 18.16.0
  pnpm: 8.8.0
Relevant Packages:
  next: 13.4.12
  eslint-config-next: 13.1.5
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 4.9.4
```

**Comment by lancejpollard** [comment]
Just updated to Next.js 14 canary.1, and so far here are my production logs:

```
 
(node:188) ExperimentalWarning: Import assertions are not a stable feature of the JavaScript language. Avoid relying on their current behavior and syntax as those might change in a future version of Node.js.
(Use `node --trace-warnings ...` to show where the warning was created)
(node:188) ExperimentalWarning: Importing JSON modules is an experimental feature and might change at any time
   Linting and checking validity of types ...
   ▲ Next.js 14.0.2-canary.1

   Creating an optimized production build ...
 ⚠ Custom webpack configuration is detected. When using a custom webpack configuration, the Webpack build worker is disabled by default. To force enable it, set the "experimental.webpackBuildWorker" option to "true". Read more: https://nextjs.org/docs/messages/webpack-build-worker-opt-out
<w> [webpack.cache.PackFileCacheStrategy] Serializing big strings (836kiB) impacts deserialization performance (consider using Buffer instead and decode when needed)
```

Still taking a while. Wondering how I can better debug what it's spending time on.

<img width="283" alt="Screenshot 2023-11-01 at 2 43 56 AM" src="https://github.com/vercel/next.js/assets/28718/d4518249-4d77-48e7-b3f4-22b60bb678a0">

**Comment by EnzoPV25** [comment]
Hi, did you find any track on how to debug this ? I also have the warning in development environment and I'd like to debbug and fix it !

**Reply by anaclumos** [reply]
Still waiting for an update

**Comment by rimzzlabs** [comment]
Im ditching nextjs now I've had enough headache just trying to make things work when building the app, thanks for 3 years 🤣

**Comment by mekhushi** [comment]
I totally get the frustration! One thing that helped me was using next-transpile-modules and tweaking Webpack configs manually. Sometimes removing unnecessary large dependencies also improves build performance. Maybe worth a shot before switching?

**Comment by secret-username** [comment]
The same issue came up on one of my project this week. Locally; deleting the .next folder and re-running build got rid of this warning and others.

**Comment by IzatIT** [comment]
+

**Comment by engripaye** [comment]
---

# ✅ **Valid Answer**

The message:

```
[webpack.cache.PackFileCacheStrategy] Serializing big strings (834kiB) impacts deserialization performance
```

comes directly from **Webpack 5** and not from your application code.
It means that **something in your build output contains a very large string**, and Webpack is trying to cache it.
When that "string" is > 500kb, Webpack warns you because:

* it makes Webpack’s filesystem cache slow
* it increases both **serialization** (writing cache) and **deserialization** (reading cache) time
* repeated builds (like Vercel) become slow

This is why your `next build` is taking ~5 minutes.

---

# ✔️ **What usually causes it**

In Next.js, these are the most common culprits:

### **1. Large JSON files imported directly in components**

Example:

```ts
import data from "@/data/bigfile.json";
```

### **2. Large `.svg` files imported as React components**

(SVG becomes a giant string literal inside the JS bundle.)

### **3. Very large static data objects stored in TS/JS files**

Example:

```ts
export const hugeData = { ... } // 1+ MB object
```

### **4. Huge environment variables encoded into the build**

Next.js embeds `NEXT_PUBLIC_*` into the client bundle.

---

# ✔️ How to debug (the straightforward way)

### **1. Enable webpack profiling**

Run:

```bash
NEXT_PRIVATE_BUILD_WORKER=1 next build --profile
```

Then inspect the output:

```bash
.next/trace
```

---

### **2. Check which files are large in the output**

Run:

```bash
du -h .next/cache/webpack | sort -hr | head
```

The biggest cached packfile usually reveals the problematic module.

---

### **3. Enable Webpack stats to get the exact module**

Add a `next.config.js` snippet:

```js
module.exports = {
  webpack(config) {
    config.stats = {
      all: false,
      assets: true,
      modules: true,
      reasons: true,
      chunks: true,
      chunkModules: true,
      errorDetails: true,
    };
    return config;
  },
};
```

Then run:

```bash
next build > build-log.txt
```

Open the file and search for modules that are > 500kb.

---

# ✔️ How to fix (once you identify the cause)

### **1. If it’s a huge JSON file**

Don’t import it directly.

Instead:

* Convert it into `.json` in `/public`, then fetch it:

```ts
const res = await fetch("/bigfile.json");
```

### **2. If it’s a large SVG**

Store it in `/public` and use:

```tsx
<img src="/large.svg" />
```

### **3. If it’s static TS/JS data**

Move it into:

* `/public/data.json`
* or fetch from an API at build/runtime

### **4. If it’s a huge NEXT_PUBLIC_ variable**

Move it to a **server-only** env variable, not exposed to the client.

---

# 🎉 **Bottom line**

* You aren’t doing anything “wrong” — Webpack is caching something very large.
* Use the profiling steps above to find the exact module.
* Move large imports out of the JS bundle (public folder, API calls, etc.).
* This **removes the warning** and **cuts build time dramatically**.

---
